### PR TITLE
Implement public API to retrieve all root items for a specific Token

### LINKED ISSUE #88: Add API to retrieve a collection's base item
An example:

`seq(a, b, c)` will create three `ParseGraph` objects when successfully parsed:

1. Containing `c`,
2. Containing `b` with `c` as a child,
3. Containing `a` with `b` as a child.

It is often desirable (e.g., in callbacks or when processing complete results) to only receive the item from 3, which contains the entire collection. An API should be created that makes this possible.

For example, `ByToken.getRoot(token)` or `ByToken.getBase(token)`.

### PR DESCRIPTION
**Pull request by jvdb**
Resolves #88.

### CODE REVIEW

File: core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java [hunk 1]
```diff
@@ -78,5 +78,18 @@ private static ParseValueList getAllValuesRecursive(final ParseGraph graph, fina
...
     }
+    
+    public static ParseItemList getAllRoots(final ParseGraph graph, final Token definition) {
+        checkNotNull(graph, "graph");
+        checkNotNull(definition, "definition");
+        return getAllRootsRecursive(graph, definition);
+    }
 
+    private static ParseItemList getAllRootsRecursive(final ParseItem item, final Token definition) {
+        if (item.getDefinition() == definition) { return ParseItemList.create(item); }
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java]: We've discussed this off-line, but I'll also write this here for future reference: We will now miss results from self-referencing `Token`s. In that case a graph can contain another graph with the same `Token` definition that can still be considered a 'root' graph.

And by self-referencing tokens I mean either:

*a*) anonymous `Token`s that parse `this` (e.g. linked lists), or;
*b*) `Token`s that eventually `sub()` to themselves (currently implemented using *a*).

File: core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java [hunk 2]
```diff
@@ -78,5 +78,22 @@ private static ParseValueList getAllValuesRecursive(final ParseGraph graph, fina
         if (head.isGraph()) { return tailResults.add(getAllValuesRecursive(head.asGraph(), definition)); }
         return tailResults;
     }
+    
+    public static ParseItemList getAllRoots(final ParseGraph graph, final Token definition) {
+        checkNotNull(graph, "graph");
+        checkNotNull(definition, "definition");
+        return (graph.getDefinition() == definition && !graph.isEmpty() ? ParseItemList.create(graph) : ParseItemList.EMPTY)
+            .add(getAllRootsRecursive(graph.head, graph, definition))
+            .add(getAllRootsRecursive(graph.tail, graph, definition));
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java]: I like this alignment of the two `.add`s, you could do that below as well, although that might look a bit too weird?

File: core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java [hunk 3]
```diff
@@ -78,5 +78,22 @@ private static ParseValueList getAllValuesRecursive(final ParseGraph graph, fina
...
+    }
+
+    private static ParseItemList getAllRootsRecursive(final ParseItem item, final ParseGraph parent, final Token definition) {
+        final ParseItemList result = (item.getDefinition() == definition && parent.getDefinition() != definition ? ParseItemList.create(item) : ParseItemList.EMPTY);
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java]: This is a pretty long line.

An idea: we introduce a `boolean hasDefinition(Token definition)` on `ParseItem`. This line would become:

```java
final ParseItemList result = (item.hasDefinition(definition) && !parent.hasDefinition(definition) ? ParseItemList.create(item) : ParseItemList.EMPTY);
```

That's still pretty long, though... Nevermind 😉

File: core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java [hunk 4]
```diff
@@ -78,5 +78,22 @@ private static ParseValueList getAllValuesRecursive(final ParseGraph graph, fina
         if (head.isGraph()) { return tailResults.add(getAllValuesRecursive(head.asGraph(), definition)); }
         return tailResults;
     }
+    
+    public static ParseItemList getAllRoots(final ParseGraph graph, final Token definition) {
+        checkNotNull(graph, "graph");
+        checkNotNull(definition, "definition");
+        return (graph.getDefinition() == definition && !graph.isEmpty() ? ParseItemList.create(graph) : ParseItemList.EMPTY)
```
Comment (rdvdijk) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java]: Are the `!graph.isEmpty()` checks actually needed?

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java [hunk 5]
```diff
@@ -229,4 +250,35 @@ public void getSubRef() throws IOException {
...
+        assertEquals(2, seqs.size);
+        assertEquals(subSeq, seqs.head.getDefinition());
+        assertEquals(subSeq, seqs.tail.head.getDefinition());
+        final ParseValue c1 = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c1.asNumeric().intValue());
+        final ParseValue c2 = seqs.tail.head.asGraph().head.asValue();
+        assertEquals(3, c2.asNumeric().intValue());
+        assertNotEquals(seqs.head.asGraph().head, seqs.tail.head.asGraph().head);
+    }
+
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: You need test here to confirm that we can now retrieve graphs created by self-referencing tokens.

Comment (jvdb) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: Agreed! I mentioned this in the commit message :)

File: core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java (hunk 4 above)
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java]: Otherwise there are 'NullPointerException's on terminators...

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java (hunk 5 above)
Comment (jvdb) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: I've added a `Sub`-test but not one that uses self-referencing, mainly because that causes a problem: the `definition` of a `Token` contained in a custom `Token` (anonymous or not) actually doesn't point to the outside class, but the embedded `Token`, which is difficult to access and reference.

Comment (jvdb) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: Added self-referencing test using a custom `Token` implementation that exposes its inner `Token`.

File: core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java (hunk 4 above)
Comment (jvdb) [inline core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java]: Additionally, it's factored slightly differently in the final commit to deal with an edge case (bonus: full coverage!)

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java (hunk 5 above)
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: Not sure I fully understand this:

> the `definition` of a `Token` contained in a custom `Token` (anonymous or not) actually doesn't point to the outside class, but the embedded `Token`, which is difficult to access and reference

You'll have to explain that to me later, I guess that's easier to explain face-to-face.

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java [hunk 6]
```diff
@@ -229,4 +252,93 @@ public void getSubRef() throws IOException {
...
+        assertEquals(smallSeq, seqs.tail.head.getDefinition());
+        final ParseValue c1 = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c1.asNumeric().intValue());
+        final ParseValue c2 = seqs.tail.head.asGraph().head.asValue();
+        assertEquals(3, c2.asNumeric().intValue());
+        assertNotEquals(seqs.head.asGraph().head, seqs.tail.head.asGraph().head);
+    }
+
+    private HashSet<ParseItem> makeSet(ParseItemList seqs) {
+        final HashSet<ParseItem> items = new HashSet<>();
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: Interface on the left, implementation on the right.

```java
final Set<ParseItem> items = new HashSet<>();
```

Also make this method return `Set<ParseItem>` and make the `seqs` argument `final`.

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java [hunk 7]
```diff
@@ -229,4 +252,93 @@ public void getSubRef() throws IOException {
...
+        for (ParseItemList current = seqs; current != null && !current.isEmpty(); current = current.tail) {
+            items.add(current.head);
+        }
+        return items;
+    }
+
+    @Test
+    public void getAllRootsMultiSub() throws IOException {
+        final ParseResult result = rep(seq(smallSeq, sub(smallSeq, currentOffset))).parse(stream(1, 2, 1, 2, 1, 2, 1, 2), enc());
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: Neat 😉

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java [hunk 8]
```diff
@@ -229,4 +252,93 @@ public void getSubRef() throws IOException {
...
+        assertEquals(3, c1.asNumeric().intValue());
+        final ParseValue c2 = seqs.tail.head.asGraph().head.asValue();
+        assertEquals(3, c2.asNumeric().intValue());
+        assertNotEquals(seqs.head.asGraph().head, seqs.tail.head.asGraph().head);
+    }
+
+    private HashSet<ParseItem> makeSet(ParseItemList seqs) {
+        final HashSet<ParseItem> items = new HashSet<>();
+        for (ParseItemList current = seqs; current != null && !current.isEmpty(); current = current.tail) {
+            items.add(current.head);
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: I don't think that any of our `ParseItem`s implement a `hashCode()`, are you sure this actually works as expected?

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java [hunk 9]
```diff
@@ -229,4 +252,93 @@ public void getSubRef() throws IOException {
...
+        final ParseResult result = rep(seq(smallSeq, sub(smallSeq, currentOffset))).parse(stream(1, 2, 1, 2, 1, 2, 1, 2), enc());
+                                                                                           /* 1: +--------+
+                                                                                           /* 2:       +--------+
+                                                                                           /* 3:             +--------+ */
+        assertTrue(result.succeeded);
+        final ParseItemList seqs = ByToken.getAllRoots(result.environment.order, smallSeq);
+        assertEquals(6, seqs.size); // Three regular and three subs.
+        final HashSet<ParseItem> items = makeSet(seqs);
+        assertEquals(seqs.size, items.size()); // Check that there are no duplicate results.
+        for (ParseItem item : items) {
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: `final ParseItem`

File: core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java [hunk 10]
```diff
@@ -229,4 +252,93 @@ public void getSubRef() throws IOException {
...
+        public CustomToken() {
+            super("", enc());
+            token = seq(any("a"), opt(this));
+        }
+
+        @Override
+        protected ParseResult parseImpl(String scope, Environment env, Encoding enc) throws IOException {
+            return token.parse(scope, env, enc);
+        }
+    }
```
Comment (rdvdijk) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: I think I now see your point. The `CustomToken` instance itself never ends up as a `definition` of any `ParseItem` in the graph, but the nested `token` does.

Comment (jvdb) [inline core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java]: Indeed!
